language/diem-vm/src/adapter_common.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{counters::*, create_access_path, data_cache::StateViewCache};
+use anyhow::Result;
+use diem_state_view::StateView;
+use diem_types::{
+    account_address::AccountAddress,
+    account_config::{self, CurrencyInfoResource, RoleId},
+    transaction::{
+        GovernanceRole, SignatureCheckedTransaction, SignedTransaction, VMValidatorResult,
+    },
+    vm_status::{StatusCode, VMStatus},
+};
+use move_core_types::{
+    identifier::{IdentStr, Identifier},
+    move_resource::MoveStructType,
+};
+use move_vm_runtime::{data_cache::MoveStorage, session::Session};
+
+use crate::logging::AdapterLogSchema;
+
+/// This trait describes the VM adapter's interface.
+pub trait VMAdapter {
+    /// Creates a new Session backed by the given storage.
+    /// TODO: this probably doesn't belong in this trait. We should be able to remove
+    /// this after refactoring Diem transaction executor code.
+    fn new_session<'r, R: MoveStorage>(&self, remote: &'r R) -> Session<'r, '_, R>;
+
+    /// Checks the signature of the given signed transaction and returns
+    /// `Ok(SignatureCheckedTransaction)` if the signature is valid.
+    fn check_signature(txn: SignedTransaction) -> Result<SignatureCheckedTransaction>;
+
+    /// Returns true if multi agent transactions are allowed.
+    fn allows_multi_agent(&self) -> Result<bool, VMStatus>;
```

Comment:
should we have a more general flag or config system? Not sure how many other settings we  forsee